File: grafana/tabs/chrony.py
"""
Chrony tab — covers all opnsense_chrony_* metrics.

Gated on the presence of opnsense_chrony_stratum so the tab only shows
when the chrony plugin is installed and scraping.

Note on label churn: with a `pool` directive chrony rotates resolved
servers over time, so the `source` label set churns. Per-source panels
use instant-query tables rather than long-range per-source timeseries
to handle this gracefully.

Rows:
  1. Service & Sync      — service_running stat, stratum stat, leap_status stat
  2. Tracking            — system_time_offset, last_offset, rms_offset, frequency gauges/ts;
                           update_interval, root_delay/dispersion; tracking_info table
  3. Sources             — sources stat; source selected/stratum/reachability/last_rx table;
                           offset timeseries
  4. Source Statistics   — source_offset_stddev, source_samples table
"""

from builder import Builder, sel, RUNSTOP


# Leap status mapping: 0=Normal, 1=Insert, 2=Delete, 3=Not synchronised
_LEAP = {
    "0": ("Normal", "green"),
    "1": ("Insert second", "yellow"),
    "2": ("Delete second", "yellow"),
    "3": ("Not synchronised", "red"),
}


def build(b: Builder):
    # ---- Sentinels ---------------------------------------------------------
    b.sentinel("has_chrony", metric="opnsense_chrony_stratum")

    # ================================================================
    # Row 1: Service & Sync
    # ================================================================
    svc = b.stat(
        "Chrony Service",
        sel("opnsense_chrony_service_running"),
        mappings=RUNSTOP, color_mode="background",
        thresholds=[{"color": "red", "value": None}, {"color": "green", "value": 1}],
        w=4, h=4,
        desc="1 = chrony service is running, 0 = stopped/disabled.",
    )
    stratum = b.stat(
        "Stratum",
        sel("opnsense_chrony_stratum"),
        unit="short", w=4, h=4,
        thresholds=[
            {"color": "green", "value": None},
            {"color": "yellow", "value": 4},
            {"color": "red", "value": 8},
        ],
        color_mode="background",
        desc=(
            "Current NTP stratum. Lower is better (1 = GPS/atomic, "
            "16 = unsynchronised)."
        ),
    )
    leap = b.stat(
        "Leap Status",
        sel("opnsense_chrony_leap_status"),
        mappings=_LEAP, color_mode="background",
        # The mapping owns the colour for 0-3; the fallback must not be green, or an
        # unmapped leap value would render as healthy (#514).
        thresholds=[{"color": "red", "value": None}],
        w=4, h=4,
        desc=(
            "chronyc Leap Status: 0=Normal, 1=Insert second, "
            "2=Delete second, 3=Not synchronised."
        ),
    )
    sources_total = b.stat(
        "Sources",
        sel("opnsense_chrony_sources"),
        unit="short", w=4, h=4,
        desc="Total number of NTP sources configured in chrony.",
    )

    # ================================================================
    # Row 2: Tracking
    # ================================================================
    tracking_info = b.table(
        "Chrony Tracking Info",
        [sel("opnsense_chrony_tracking_info")],
        w=12, h=4,
        excludes=["Value", "__name__", "job", "instance"],
        renames={
            "reference_id": "Reference ID",
            "reference_name": "Reference Name",
            "leap_status": "Leap Status",
        },
        sort_by="Reference Name",
        desc="Current chrony reference info (value always 1).",
    )
    sys_offset = b.ts(
        "System Time Offset",
        [(sel("opnsense_chrony_system_time_offset_seconds"), "offset (fast=+, slow=−)"),
         (sel("opnsense_chrony_last_offset_seconds"), "last offset"),
         (sel("opnsense_chrony_rms_offset_seconds"), "RMS offset")],
        # Signed: fast = positive, slow = negative. The builder's min0 default pinned the
        # axis floor at 0 and drew every slow clock below the visible area, so half of all
        # drift incidents looked identical to a perfectly synced clock (#512).
        unit="s", min0=False, w=12, h=8,
        desc=(
            "System clock offset from NTP time. Negative = slow; "
            "rms_offset is the long-term root-mean-square."
        ),
    )
    freq_ts = b.ts(
        "Frequency Error (ppm)",
        [(sel("opnsense_chrony_frequency_ppm"), "frequency"),
         (sel("opnsense_chrony_residual_frequency_ppm"), "residual"),
         (sel("opnsense_chrony_skew_ppm"), "skew")],
        # frequency and residual are signed; only skew is non-negative (#512).
        unit="ppm", min0=False, w=12, h=8,
        desc="chrony frequency error metrics in parts per million.",
    )
    root_ts = b.ts(
        "Root Delay & Dispersion",
        [(sel("opnsense_chrony_root_delay_seconds"), "root delay"),
         (sel("opnsense_chrony_root_dispersion_seconds"), "root dispersion")],
        unit="s", w=12, h=8,
        desc=(
            "Root delay = total round-trip to the reference clock. "
            "Root dispersion = accumulated error from the reference."
        ),
    )
    update_interval = b.ts(
        "Tracking Update Interval",
        [(sel("opnsense_chrony_update_interval_seconds"), "update interval")],
        unit="s", w=12, h=8,
        desc="Interval between the last two clock updates.",
    )

    # ================================================================
    # Row 3: Sources
    # ================================================================
    source_table = b.table(
        "Source Status",
        [
            sel("opnsense_chrony_source_selected"),
            sel("opnsense_chrony_source_stratum"),
            sel("opnsense_chrony_source_reachability"),
            sel("opnsense_chrony_source_last_rx_seconds"),
        ],
        w=24, h=10,
        excludes=["__name__", "job", "instance"],
        renames={
            "source": "Source",
            "mode": "Mode",
            "Value #A": "Selected",
            "Value #B": "Stratum",
            "Value #C": "Reachability",
            "Value #D": "Last RX (s)",
        },
        unit_overrides={"Value #D": "s"},
        sort_by="Source",
        desc=(
            "Per-source NTP status. Selected=1 means this is the active reference. "
            "Reachability is an 8-bit shift register (255=fully reachable). "
            "Last RX = seconds since the last valid packet was received."
        ),
    )
    source_offset = b.ts(
        "Source Offset",
        [(sel("opnsense_chrony_source_offset_seconds"), "{{source}}")],
        unit="s", min0=False, w=24, h=8,   # signed offset per source (#512)
        desc=(
            "Measured time offset from each NTP source. "
            "Note: source label set may churn with a pool directive."
        ),
    )

    # ================================================================
    # Row 4: Source Statistics
    # ================================================================
    source_stats_table = b.table(
        "Source Statistics",
        [
            sel("opnsense_chrony_source_offset_stddev_seconds"),
            sel("opnsense_chrony_source_samples"),
        ],
        w=24, h=8,
        excludes=["__name__", "job", "instance"],
        renames={
            "source": "Source",
            "Value #A": "Offset Std Dev (s)",
            "Value #B": "Samples",
        },
        unit_overrides={"Value #A": "s"},
        sort_by="Source",
        desc=(
            "Per-source regression statistics from chronyc sourcestats. "
            "Std Dev measures dispersion of recent offset samples."
        ),
    )

    # ================================================================
    # Tab assembly
    # ================================================================
    b.tab("Chrony", [
        b.autogrid_row("Service & Sync",
              [svc, stratum, leap, sources_total],
              present="has_chrony"),
        b.row("Tracking",
              [tracking_info, sys_offset, freq_ts, root_ts, update_interval],
              present="has_chrony"),
        b.row("Sources",
              [source_table, source_offset],
              present="has_chrony"),
        b.row("Source Statistics",
              [source_stats_table],
              present="has_chrony"),
    ])

File: internal/collector/chrony.go
package collector

import (
	"context"
	"log/slog"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/rknightion/opnsense2otel/v5/opnsense"
)

type chronyCollector struct {
	log *slog.Logger

	serviceRunning       *prometheus.Desc
	stratum              *prometheus.Desc
	systemTimeOffset     *prometheus.Desc
	lastOffset           *prometheus.Desc
	rmsOffset            *prometheus.Desc
	frequencyPPM         *prometheus.Desc
	residualFrequencyPPM *prometheus.Desc
	skewPPM              *prometheus.Desc
	rootDelay            *prometheus.Desc
	rootDispersion       *prometheus.Desc
	updateInterval       *prometheus.Desc
	leapStatus           *prometheus.Desc
	trackingInfo         *prometheus.Desc
	sourcesTotal         *prometheus.Desc
	sourceSelected       *prometheus.Desc
	sourceStratum        *prometheus.Desc
	sourceReachability   *prometheus.Desc
	sourceLastRx         *prometheus.Desc
	sourceOffset         *prometheus.Desc
	sourceOffsetStdDev   *prometheus.Desc
	sourceSamples        *prometheus.Desc

	subsystem string
	instance  string
}

func init() {
	collectorInstances = append(collectorInstances, &chronyCollector{
		subsystem: ChronySubsystem,
	})
}

func (c *chronyCollector) Name() string { return c.subsystem }

func (c *chronyCollector) Register(namespace, instanceLabel string, log *slog.Logger) {
	c.log = log
	c.instance = instanceLabel
	c.log.Debug("Registering collector", "collector", c.Name())

	srcLabels := []string{"source", "mode"}

	c.serviceRunning = buildPrometheusDesc(c.subsystem, "service_running",
		"Whether the Chrony service is running (1 = running, 0 = stopped/disabled)", nil)
	c.stratum = buildPrometheusDesc(c.subsystem, "stratum",
		"Stratum of the NTP source chrony is currently tracking", nil)
	c.systemTimeOffset = buildPrometheusDesc(c.subsystem, "system_time_offset_seconds",
		"Current system time offset relative to NTP time in seconds (fast = positive, slow = negative)", nil)
	c.lastOffset = buildPrometheusDesc(c.subsystem, "last_offset_seconds",
		"Estimated local offset on the last clock update in seconds", nil)
	c.rmsOffset = buildPrometheusDesc(c.subsystem, "rms_offset_seconds",
		"Long-term average of the offset values in seconds", nil)
	c.frequencyPPM = buildPrometheusDesc(c.subsystem, "frequency_ppm",
		"Rate at which the system clock would be wrong if chrony was not correcting it (ppm)", nil)
	c.residualFrequencyPPM = buildPrometheusDesc(c.subsystem, "residual_frequency_ppm",
		"Residual frequency offset of the NTP source (ppm)", nil)
	c.skewPPM = buildPrometheusDesc(c.subsystem, "skew_ppm",
		"Estimated error bound on the frequency (ppm)", nil)
	c.rootDelay = buildPrometheusDesc(c.subsystem, "root_delay_seconds",
		"Total of the network path delays to the stratum-1 NTP server (seconds)", nil)
	c.rootDispersion = buildPrometheusDesc(c.subsystem, "root_dispersion_seconds",
		"Total dispersion accumulated through all NTP servers back to stratum-1 (seconds)", nil)
	c.updateInterval = buildPrometheusDesc(c.subsystem, "update_interval_seconds",
		"Interval between the last two clock updates (seconds)", nil)
	c.leapStatus = buildPrometheusDesc(c.subsystem, "leap_status",
		"Leap second status (0=Normal, 1=Insert second, 2=Delete second, 3=Not synchronised)", nil)
	c.trackingInfo = buildPrometheusDesc(c.subsystem, "tracking_info",
		"Chrony tracking info label gauge (always 1) - pool-rotation note: source labels churn with pool directives; use instant queries/tables",
		[]string{"reference_id", "reference_name", "leap_status"})
	c.sourcesTotal = buildPrometheusDesc(c.subsystem, "sources",
		"Current number of NTP sources chrony is currently tracking", nil)
	c.sourceSelected = buildPrometheusDesc(c.subsystem, "source_selected",
		"Whether this source is currently selected as the best available (1 = selected)",
		srcLabels)
	c.sourceStratum = buildPrometheusDesc(c.subsystem, "source_stratum",
		"Stratum of this NTP source", srcLabels)
	c.sourceReachability = buildPrometheusDesc(c.subsystem, "source_reachability",
		"Source reachability register (0–255; 255 means all 8 last polls succeeded)", srcLabels)
	c.sourceLastRx = buildPrometheusDesc(c.subsystem, "source_last_rx_seconds",
		"Time since the last packet was received from this source (seconds)", srcLabels)
	c.sourceOffset = buildPrometheusDesc(c.subsystem, "source_offset_seconds",
		"Measured offset of the source (seconds)", srcLabels)
	c.sourceOffsetStdDev = buildPrometheusDesc(c.subsystem, "source_offset_stddev_seconds",
		"Estimated standard deviation of the source offset (seconds)", []string{"source"})
	c.sourceSamples = buildPrometheusDesc(c.subsystem, "source_samples",
		"Number of samples currently held in the regression (sourcestats)", []string{"source"})
}

func (c *chronyCollector) Describe(ch chan<- *prometheus.Desc) {
	for _, d := range []*prometheus.Desc{
		c.serviceRunning, c.stratum, c.systemTimeOffset, c.lastOffset, c.rmsOffset,
		c.frequencyPPM, c.residualFrequencyPPM, c.skewPPM, c.rootDelay, c.rootDispersion,
		c.updateInterval, c.leapStatus, c.trackingInfo, c.sourcesTotal,
		c.sourceSelected, c.sourceStratum, c.sourceReachability, c.sourceLastRx,
		c.sourceOffset, c.sourceOffsetStdDev, c.sourceSamples,
	} {
		ch <- d
	}
}

func (c *chronyCollector) Update(_ context.Context, client *opnsense.Client, ch chan<- prometheus.Metric) *opnsense.APICallError {
	data, err := client.FetchChronyStatus()
	if err != nil {
		return err
	}

	// Plugin absent (tracking 404): stay completely silent AND skip the
	// service-status probe — the service endpoint would just 404 too.
	if !data.Present {
		return nil
	}

	// Fetch service status first so it's emitted even when chronyd is stopped.
	status, present, sErr := client.FetchServiceStatusOptional("chronyServiceStatus")
	if sErr != nil {
		c.log.Warn("failed to fetch chrony service status", "err", sErr)
	} else if present {
		running := 0.0
		if status == "running" {
			running = 1.0
		}
		ch <- prometheus.MustNewConstMetric(c.serviceRunning, prometheus.GaugeValue,
			running, c.instance)
	}

	// Tracking metrics — only when chronyd is running and has a valid source.
	if data.Tracking.Valid {
		tr := data.Tracking
		ch <- prometheus.MustNewConstMetric(c.stratum, prometheus.GaugeValue, tr.Stratum, c.instance)
		ch <- prometheus.MustNewConstMetric(c.systemTimeOffset, prometheus.GaugeValue, tr.SystemTimeOffsetSeconds, c.instance)
		ch <- prometheus.MustNewConstMetric(c.lastOffset, prometheus.GaugeValue, tr.LastOffsetSeconds, c.instance)
		ch <- prometheus.MustNewConstMetric(c.rmsOffset, prometheus.GaugeValue, tr.RMSOffsetSeconds, c.instance)
		ch <- prometheus.MustNewConstMetric(c.frequencyPPM, prometheus.GaugeValue, tr.FrequencyPPM, c.instance)
		ch <- prometheus.MustNewConstMetric(c.residualFrequencyPPM, prometheus.GaugeValue, tr.ResidualFrequencyPPM, c.instance)
		ch <- prometheus.MustNewConstMetric(c.skewPPM, prometheus.GaugeValue, tr.SkewPPM, c.instance)
		ch <- prometheus.MustNewConstMetric(c.rootDelay, prometheus.GaugeValue, tr.RootDelaySeconds, c.instance)
		ch <- prometheus.MustNewConstMetric(c.rootDispersion, prometheus.GaugeValue, tr.RootDispersionSeconds, c.instance)
		ch <- prometheus.MustNewConstMetric(c.updateInterval, prometheus.GaugeValue, tr.UpdateIntervalSeconds, c.instance)
		ch <- prometheus.MustNewConstMetric(c.leapStatus, prometheus.GaugeValue, tr.LeapStatusValue, c.instance)
		ch <- prometheus.MustNewConstMetric(c.trackingInfo, prometheus.GaugeValue, 1,
			tr.ReferenceID, tr.ReferenceName, tr.LeapStatus, c.instance)
	}

	// Sources total + per-source metrics. Skip entirely when the sources sub-fetch
	// failed, so a transient error is not reported as a real sources=0 that
	// looks like chronyd lost all its NTP sources (#163).
	if data.HasSources {
		ch <- prometheus.MustNewConstMetric(c.sourcesTotal, prometheus.GaugeValue,
			float64(len(data.Sources)), c.instance)

		for _, src := range data.Sources {
			selected := 0.0
			if src.State == "*" {
				selected = 1.0
			}
			ch <- prometheus.MustNewConstMetric(c.sourceSelected, prometheus.GaugeValue,
				selected, src.Name, src.Mode, c.instance)
			ch <- prometheus.MustNewConstMetric(c.sourceStratum, prometheus.GaugeValue,
				src.Stratum, src.Name, src.Mode, c.instance)
			ch <- prometheus.MustNewConstMetric(c.sourceReachability, prometheus.GaugeValue,
				src.Reachability, src.Name, src.Mode, c.instance)
			if src.HasLastRx {
				ch <- prometheus.MustNewConstMetric(c.sourceLastRx, prometheus.GaugeValue,
					src.LastRxSeconds, src.Name, src.Mode, c.instance)
			}
			if src.HasOffset {
				ch <- prometheus.MustNewConstMetric(c.sourceOffset, prometheus.GaugeValue,
					src.OffsetSeconds, src.Name, src.Mode, c.instance)
			}
		}
	}

	// Per-source stats (stddev + samples). Skipped when the sourcestats sub-fetch
	// failed (#163).
	if data.HasSourceStats {
		for _, st := range data.SourceStats {
			if st.HasStdDev {
				ch <- prometheus.MustNewConstMetric(c.sourceOffsetStdDev, prometheus.GaugeValue,
					st.StdDevSeconds, st.Name, c.instance)
			}
			ch <- prometheus.MustNewConstMetric(c.sourceSamples, prometheus.GaugeValue,
				st.Samples, st.Name, c.instance)
		}
	}

	return nil
}
